[2단계 - 로또 리팩토링] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
안녕하세요 로키, 2단계 미션 제출합니다.😃
2단계에서는 마지막으로 주신 피드백들을 주로 반영하였습니다. 

### 주요 변경사항은 다음과 같습니다.
- domain 패키지를 생성하여 핵심 도메인 로직이라고 여겨지는 객체들을 보관했습니다.
- LottoManager의 기능을 LottoMachine에 위임했습니다.
- 도메인 객체의 public 메서드에 대해 누락되었던 단위 테스트 수행했습니다.
- 랜덤 값을 제어하고 테스트의 멱등성을 지키기 위해  랜덤과 관련한 로직을 인터페이스화했습니다.(LottoNumberGenerator)
- 테스트 메서드를 더 파악하기 쉽게끔 given/when/then 스타일을 적용했습니다.
- 1단계의 피드백을 반영해 코드 수정 등

### 추가로 궁금한 점은 다음과 같습니다.
- 2단계 미션의 핵심은 MVC 패턴에 기반하여 리팩토링하는 것이었는데요, 제가 학습하고 이해한 MVC 패턴 내 Controller의 역할은 다음과 같습니다. `도메인이 뷰에 의존하지 않게끔 Controller는 도메인과 뷰 모두에 의존하며 둘 사이의 로직을 수행한다.` 이렇게 보았을 때 현재 코드의 Application main()메서드도 충분히 Controller의 역할을 잘 수행하고 있다고 판단하여 별도의 Controller를 생성하지 않았습니다. 이에 대해 로키는 어떻게 생각하시는 지 궁금합니다!

kjyyjk committed:
refactor: Rank 로직 리팩토링

- 당첨금을 파악하기 쉽게 세자리마다 `_`를 추가한다.

- 기존 checkRank() 로직은 for문 안에 if문이 존재해 depth 1단계를 넘긴다.
따라서 메서드내 들여쓰기가 한단계만 존재하도록 수정한다.

- 기존 코드는 calculateAllWinningAmount() 내부에서 winningAmount를 getter로
가져오고 있다. winningAmount는 Rank 자신의 상태이기 때문에 바로 접근하도록
수정한다.

kjyyjk committed:
refactor: entrySet()을 활용한 리팩토링

- 기존 WinningStatistics의 calculateTotalWinningAmount()에서는
총 당첨금액을 계산하기 위해 total 변수에다가 각 rank별 당첨 금액을 합해 반환한다.
map.keySet()을 활용해 key값인 rank를 기반으로 로직을 수행했으며, 해당 key에 대한
value를 받아오기 위해 map에서 다시 key를 기반으로 get하는 것도 확인할 수 있다.
이를 map.entrySet()을 활용하게끔 리팩토링한다.
entrySet()은 map에 포함된 매핑(Entry)들의 집합을 반환한다.
각 집합 내 entry에서 바로 key와 value에 접근할 수 있다.
entry 집합과 stream을 활용해 비교적 간단하게 총 당첨금액을 반환할 수 있다.

kjyyjk committed:
test: WinningNumbers의 contains() 단위 테스트

kjyyjk committed:
refactor: WinningNumbers를 Lotto로 대체

- WinningNumbers도 결국은 6개의 고유한 번호를 가진 Lotto와도 같다.
따라서 WinningNumbers를 Lotto로 대체한다.
이 과정에서 WinningNumbers 내부에 있던 검증 로직들을 Lotto로 옮긴다.

kjyyjk committed:
test: 발행된 로또의 범위가 올바른지 추가 검증

kjyyjk committed:
test: Lotto 번호 검증 로직의 테스트 방식 변경

- 기존에는 Lotto의 validateLottoNumber()를 직접 호출하여 검증하고 있었으나
Lotto를 생성할 때 적절하게 예외를 던지는 지 테스트하는 방식으로 수정한다.

kjyyjk committed:
refactor: 잘못된 구매금액 입력 시 잘못된 값 함께 출력

kjyyjk committed:
style: 불필요한 변수 선언 제거

kjyyjk committed:
test: WinningStatistics의 수익률 계산 테스트

kjyyjk committed:
refactor: domain 패키지 생성

- 기존에는 UI 관련한 객체들을 모아놓는 view패키지가 존재했다.
이번에는 핵심 비즈니스 로직을 가지는 도메인 객체들을 모아놓는
domain 패키지를 생성한다.

kjyyjk committed:
refactor: 난수 로또 번호 생성로직 인터페이스화

- 기존 코드에서는 테스트마다 생성되는 로또 번호가 매번 달라지기 때문에
테스트의 멱등성이 깨진다. 랜덤값은 임의로 제어하기가 쉽지 않기 때문에
인터페이스화하여 로직을 분리하고, 테스트용 generator를 만들어 사용한다.

kjyyjk committed:
refactor: LottoManager의 기능을 LottoMachine에 위임

- 기존에는 현실 세계를 반영하여 LottoManger를 두었다.
하지만 LottoManager에 별다른 로직이 존재하지 않기도 하고,
LottoMachine이 구입금액에 따른 로또들을 발행하는 것도 적절하다고 판단됐다.

kjyyjk committed:
test: WinningLotto의 로또 당첨 통계 테스트

kjyyjk committed:
style: test given/when/then 스타일 적용

Rok93 commented on src/main/java/lotto/Application.java:
```diff
@@ -12,14 +17,15 @@ public static void main(String[] args) {
         WinningLotto winningLotto = getWinningLotto();
         WinningStatistics winningStatistics = winningLotto.calculateStatistics(lottos);
-        double returnRate = winningStatistics.calculateReturnRate(lottos.size() * LottoManager.LOTTO_UNIT_PRICE);
+        double returnRate = winningStatistics.calculateReturnRate(lottos.size() * LottoMachine.LOTTO_UNIT_PRICE);
```
질문에 대한 답변은 여기에 남겨둘게요. 😃
> 2단계 미션의 핵심은 MVC 패턴에 기반하여 리팩토링하는 것이었는데요, 제가 학습하고 이해한 MVC 패턴 내 Controller의 역할은 다음과 같습니다. 도메인이 뷰에 의존하지 않게끔 Controller는 도메인과 뷰 모두에 의존하며 둘 사이의 로직을 수행한다. 이렇게 보았을 때 현재 코드의 Application main()메서드도 충분히 Controller의 역할을 잘 수행하고 있다고 판단하여 별도의 Controller를 생성하지 않았습니다. 이에 대해 로키는 어떻게 생각하시는 지 궁금합니다!

레오의 의견에 동의해요. 😃
Application이 충분히 Controller의 역할을 하고있다고 생각해요. 

단지 별도의 controller 패키지와 Controller 클래스를 따로 두게되면 패키지 구조를 명시적으로 보여줄 수 있을 것 같은데요.
함께 일하는 팀원들이 있다면, 팀의 생산성을 높이는데 도움이 많이될 것 같아요. 🤔

kjyyjk replied:
다음 페어 프로그래밍에서는 페어와 의논 후 controller 패키지, Controller 클래스를 도입해보는 방향으로 해봐야겠어요😃 감사합니다

Rok93 commented on src/main/java/lotto/domain/SystemLottoNumberGenerator.java:
```diff
@@ -0,0 +1,21 @@
+    public List<Integer> generate() {
+        Set<Integer> numbers = new HashSet<>();
+        while (numbers.size() < LOTTO_SIZE) {
+            numbers.add(new Random().nextInt(MAX_LOTTO_NUMBER - 1) + 1);
```
generate() 메서드를 호출할 때마다 매번 Random 객체를 생성할 것 같은데요. 

Random 객체를 상수로 추출해서 재사용해보면 어떨까요? 😃

kjyyjk replied:
반영했습니다!

Rok93 commented on src/main/java/lotto/Application.java:
```diff
@@ -12,14 +17,15 @@ public static void main(String[] args) {
         WinningLotto winningLotto = getWinningLotto();
         WinningStatistics winningStatistics = winningLotto.calculateStatistics(lottos);
-        double returnRate = winningStatistics.calculateReturnRate(lottos.size() * LottoManager.LOTTO_UNIT_PRICE);
+        double returnRate = winningStatistics.calculateReturnRate(lottos.size() * LottoMachine.LOTTO_UNIT_PRICE);
```
전체적으로 너무 깔끔하게 잘 작성해주셨네요. 👍
MVC 패턴에 관련해서는 크게 피드백 드릴 부분이 없는 것 같습니다. 💯

단지 이 부분에서 조금 아쉬운 점은  구입 금액을 다시 계산하고있는 것 같은 부분인데요. 
이미 앞에서 구입금액을 분명 입력 받았었던 것 같은데, 입력받은 구입 금액을 대신 사용해볼 수도 있을 것 같아요.

kjyyjk replied:
그게 좋을 거 같아요!
반영했습니다😃

Rok93 commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,46 @@
+    public List<Lotto> purchase(final int purchaseAmount) {
+        validatePurchaseAmount(purchaseAmount);
+        int lottoAmount = purchaseAmount / LOTTO_UNIT_PRICE;
+        List<Lotto> lottos = new ArrayList<>();
+        for (int i = 0; i < lottoAmount; i++) {
+            Lotto lotto = createLotto();
+            lottos.add(lotto);
+        }
+        return lottos;
```
아래와 같이 변경해볼 수도 있을 것 같아요. 😉
```suggestion
        return Stream.generate(this::createLotto)
            .limit(lottoAmount)
            .toList();
```

kjyyjk replied:
아직 Stream을 사용하는 것이 익숙하지 않았었는데, 의식적으로 사용할 수 있도록 연습을 해봐야겠어요.
감사합니다!

Rok93 commented on src/main/java/lotto/domain/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.List;
+
+public interface LottoNumberGenerator {
```
결국 LottoNumbers가 아닌 Lotto를 생성하기위해 이 Generator가 필요한 것으로보이는데요. 🤔 
LottoNumber 대신에 Lotto를 생성해주는 LottoGenerator를 만들어보면 어떨까요?

kjyyjk replied:
generate를 호출하는 곳이 한 곳이고 반환 받은 값으로 바로 Lotto 객체를 생성하기 때문에, 로키 말대로 LottoGenerator로 수정하는 게 좋을 거 같아요! 반영하겠습니다.

Rok93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -12,27 +12,39 @@ public class Lotto {
-        return sortedNumbers;
-    }
-
     public static void validateLottoNumber(final int lottoNumber) {
```
이 로직이 왜 public 메서드인가 싶어서 확인해봤더니 Application에서도 이 유효성 검증 로직을 사용하고 계시군요? 

로또 숫자(Integer)에 대한 유효성 검증을 여기저기서 하고있는 이유가 뭘까요? 
아마도 Integer는 로또 숫자의 비즈니스 규칙 외의 숫자들도 모두 가질 수 있기 때문에 매번 검증로직이 필요할거 같아요. 

로또 숫자를 상징하는 타입을 정의해볼 수는 없을까요? 🤔 
아래의 글을 참고해보시면 도움이 되실 것 같아요. 😉

[원시 타입을 포장](https://tecoble.techcourse.co.kr/post/2020-05-29-wrap-primitive-type/)해보면 어떨까요? 😃

kjyyjk replied:
원시 타입인 로또 번호를 LottoNumber로 포장해주니, 로또 번호에 대한 검증 로직의 책임 소재가 분명해진 것 같아 마음에 들어요! 좋은 글 소개도 감사드려요😃

Rok93 reviewed (changes requested):
레오 2단계도 잘 구현해주셨네요. 👍💯
2단계 너무 잘 구현해주셔서 바로 머지할 수도 있었지만 
조금 더 짚고 넘어가면 좋을 부분들이 보여서 추가 코멘트 남겨두었어요. 😃
확인해서 반영 부탁드릴게요. 🙏

kjyyjk committed:
refactor: LottoNumbersGenerator에서 LottoGenerator로 변경

- 기존 코드는 LottoNumbersGenerator에서 정수 리스트를 생성하여 반환한다.
하지만 이를 호출하는 곳이 한곳이고, 반환받은 정수 리스트로 바로 Lotto를
생성하고 있다. 따라서 Lotto를 생성하여 반환하게끔 수정한다.

kjyyjk committed:
refactor: Stream.generate()를 활용하여 여러장의 로또 발행

- Stream.generate()는 순서없는 무한 스트림을 반환한다.
매개변수로는 함수형 인터페이스인 Supplier를 받는다.
createLotto 호출하여 반환하는 Supplier를 작성하여 넘겼기 때문에
Lotto 무한 스트림이 생성될 것이고, limit()로 원하는 개수만큼만 가져올 수 있다.

kjyyjk committed:
refactor: Random을 상수로 분리

- 기존 코드는 generate() 호출 시마다 Random 객체를 생성하고 있다.
따라서 Random을 static final 상수로 분리하여 재사용한다.

kjyyjk committed:
refactor: 입력 받은 구입 금액을 재사용

- 기존에는 구입 금액을 입력 받았음에도 다시 계산하여 사용하고 있기 때문에,
이를 수정하였다.

kjyyjk committed:
refactor: 로또 번호를 LottoNumber로 포장

- 기존에는 로또 번호를 Integer 원시 타입으로 사용하고 있어,
로또 번호에 대한 검증 로직이 Lotto 클래스에 있었다.
그리고 이 검증 로직을 외부에서 사용하기 위해 public으로 열어두었었다.
하지만 로또 번호에 대한 검증 로직은 로또 번호 자신의 책임이라고 판단되어
LottoNumber 객체로 포장한다.

kjyyjk committed:
test: LottoNumber 검증 로직 단위 테스트

kjyyjk committed:
refactor: 테스트 패키지 이동

Rok93 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,43 @@
+
+    private int lottoNumber;
+
+    public LottoNumber(final int lottoNumber) {
```
LottoNumber의 경우 1~45 까지의 LottoNumber 객체만 사용하게되는데요. 
1 ~ 45 까지의 LottoNumber를 미리 [캐싱](https://tecoble.techcourse.co.kr/post/2020-06-24-caching-instance/)해두고 재사용해봐도 좋았을 것 같아요. 😃

Rok93 commented on src/main/java/lotto/domain/SystemLottoGenerator.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Random;
+import java.util.Set;
+
+public class SystemLottoGenerator implements LottoGenerator {
```
System이라는 Prefix를 붙이신 의도가 궁금합니다. 🤔
저라면 Random으로 로또를 생성하니 `RandomLottoGenerator`와 같이 네이밍을했을 것 같아요!

kjyyjk replied:
테스트 관련 인강에서 이런 식으로 네이밍하는걸 본 뒤로 습관적으로 System을 붙이고는 했었는데,
테스트를 위한 구현체와 다르게 난수로 구성된 로또를 생성하여 반환하니까 ```RandomLottoGenerator```가
더 자연스러운 거 같아요.😃
반영했습니다.

Rok93 commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -22,27 +22,26 @@ public WinningStatistics calculateStatistics(final List<Lotto> lottos) {
 
         for (final Lotto lotto : lottos) {
-            Rank rank = checkRank(lotto.getNumbers(), winningNumbers.getWinningNumbers(), bonusNumber);
+            Rank rank = checkRank(lotto.getNumbers(), winningNumbers.getNumbers(), bonusNumber);
```
2, 3 번째 메서드 파라미터는 어짜피 WinningLotto의 상태인데, 굳이 여기서 파라미터를 주입해줬을 필요가 잇나 싶어요. 🤔

또한 WinningLotto 안의 `checkRank`, `calculateMatchCount` 두 메서드는 불필요한 것 같아요. 
오히려 lotto 객체에게 calculateMatchCount의 메시지를 보냈다면 객체와 객체간의 협력이 조금 더 자연스러웠을 것 같아요.

```suggestion
            Rank rank = Rank.checkRank(lotto.calculateMatchCount(winningNumbers), lotto.contains(bonusNumber));
```

kjyyjk replied:
로키가 말씀해주신 방향으로 고민해보고 수정했더니 WinningLotto와 Lotto 객체 간의 협력이 자연스러워졌어요.
거기다가 기존에 private 메서드였던 calculateMatchCount를 단위 테스트할 수 있어져서 더욱 좋은 거 같아요!

Rok93 commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -21,7 +26,11 @@ class RankTest {
             delimiterString = ":")
     @ParameterizedTest
     void 당첨_개수에_따른_해당_등수의_총_당첨금을_반환한다(Rank rank, int winningCount, int expected) {
+        //given
```
ParameterizedTest를 쓸 때, `// given` 주석의 위치가 약간 모호하다고 생각하는데요. 🤔
저는 이렇게 애매한 경우는 그냥 과감히 생략하는 편이에요. 😃

kjyyjk replied:
작성할 때 긴가민가했었는데 저도 로키처럼 과감히 생략해보도록 할께요!

Rok93 reviewed (approved):
레오 피드백 반영 잘해주셨네요. 👍👍 
2단계도 충분히 잘 구현해주신 것 같아서, 로또 미션은 이만 종료하도록하겠습니다. 👏👏 
추가적으로 몇몇 코멘트 남겨두었는데, 가볍게 확인한번 부탁드릴게요. 😉
다음 미션도 화이팅입니다.  💪💪💪
